File: shared/schema/document_model.py
"""
Gateway → Engine data contract.

This file is the Pydantic mirror of document_model.ts.
Any change here MUST be reflected in document_model.ts.
Both are validated against document_model.schema.json in CI.
"""

from __future__ import annotations

from typing import Annotated, Literal, Union

from pydantic import BaseModel, Field

SCHEMA_VERSION = "1.0"

# ── Metadata ──────────────────────────────────────────────────────────────────


class Author(BaseModel):
    name: str
    affiliation_refs: list[str]
    email: str | None = None


class Affiliation(BaseModel):
    key: str                    # matched by Author.affiliation_refs
    institution: str
    department: str | None = None
    city: str | None = None
    country: str | None = None


class Reference(BaseModel):
    key: str                    # citation label, e.g. "1"
    authors: list[str]
    title: str
    venue: str | None = None    # journal or conference name
    year: int
    volume: str | None = None
    issue: str | None = None
    pages: str | None = None    # e.g. "123–130"
    doi: str | None = None
    url: str | None = None


# ── Block types ───────────────────────────────────────────────────────────────


class ParagraphBlock(BaseModel):
    type: Literal["paragraph"]
    text: str                   # may contain inline [EQ n] refs
    indent: bool


class HeadingBlock(BaseModel):
    type: Literal["heading"]
    level: Literal[1, 2, 3]
    text: str
    numbering: str | None = None   # resolved by gateway, e.g. "II.A"


class FigureBlock(BaseModel):
    type: Literal["figure", "wide_figure"]
    anchor: str                 # e.g. "FIG 1"
    image_ref: str              # path relative to gateway/storage/
    caption: str


class TableBlock(BaseModel):
    type: Literal["table", "wide_table"]
    anchor: str                 # e.g. "TABLE 1"
    caption: str
    rows: list[list[str]]       # rows[row][col] — plain cell text
    header_row: bool


class EquationBlock(BaseModel):
    type: Literal["equation"]
    anchor: str                 # e.g. "EQ 1"
    latex: str                  # passed verbatim to the math backend
    inline: bool


class ListBlock(BaseModel):
    type: Literal["list"]
    style: Literal["bullet", "numbered"]
    items: list[str]


Block = Annotated[
    Union[
        ParagraphBlock,
        HeadingBlock,
        FigureBlock,
        TableBlock,
        EquationBlock,
        ListBlock,
    ],
    Field(discriminator="type"),
]


# ── Compile options ───────────────────────────────────────────────────────────


class CompileOptions(BaseModel):
    math_backend: Literal["matplotlib", "katex", "omml"] = "matplotlib"
    page_size: Literal["letter", "a4"] = "a4"
    column_sep_pt: float | None = None  # defaults to IEEE spec value


# ── Root ──────────────────────────────────────────────────────────────────────


class DocumentMetadata(BaseModel):
    title: str
    authors: list[Author]
    affiliations: list[Affiliation]
    abstract: str
    keywords: list[str]
    conference: str | None = None


class DocumentModel(BaseModel):
    schema_version: Literal["1.0"]
    metadata: DocumentMetadata
    blocks: list[Block]
    references: list[Reference] | None = None
    compile_options: CompileOptions

File: shared/schema/document_model.ts
/**
 * Gateway → Engine data contract.
 *
 * This file is the TypeScript source of truth.
 * Any change here MUST be reflected in document_model.py.
 * Both are validated against document_model.schema.json in CI.
 */

export const SCHEMA_VERSION = '1.0' as const;

// ── Metadata ──────────────────────────────────────────────────────────────────

export interface Author {
  name: string;
  affiliation_refs: string[];   // keys into DocumentMetadata.affiliations
  email?: string;
}

export interface Affiliation {
  key: string;                  // e.g. "1", "2" — matched by Author.affiliation_refs
  institution: string;
  department?: string;
  city?: string;
  country?: string;
}

export interface Reference {
  key: string;                  // citation label, e.g. "1"
  authors: string[];
  title: string;
  venue?: string;               // journal or conference name
  year: number;
  volume?: string;
  issue?: string;
  pages?: string;               // e.g. "123–130"
  doi?: string;
  url?: string;
}

// ── Block types ───────────────────────────────────────────────────────────────

export interface ParagraphBlock {
  type: 'paragraph';
  text: string;                 // may contain inline [EQ n] refs
  indent: boolean;
}

export interface HeadingBlock {
  type: 'heading';
  level: 1 | 2 | 3;
  text: string;
  numbering?: string;           // e.g. "II", "II.A" — resolved by the gateway, not the engine
}

export interface FigureBlock {
  type: 'figure' | 'wide_figure';
  anchor: string;               // e.g. "FIG 1"
  image_ref: string;            // path relative to gateway/storage/
  caption: string;
}

export interface TableBlock {
  type: 'table' | 'wide_table';
  anchor: string;               // e.g. "TABLE 1"
  caption: string;
  rows: string[][];             // rows[row][col] — plain cell text
  header_row: boolean;
}

export interface EquationBlock {
  type: 'equation';
  anchor: string;               // e.g. "EQ 1"
  latex: string;                // passed verbatim to the math backend
  inline: boolean;
}

export interface ListBlock {
  type: 'list';
  style: 'bullet' | 'numbered';
  items: string[];
}

export type Block =
  | ParagraphBlock
  | HeadingBlock
  | FigureBlock
  | TableBlock
  | EquationBlock
  | ListBlock;

// ── Compile options ───────────────────────────────────────────────────────────

export interface CompileOptions {
  math_backend: 'matplotlib' | 'katex' | 'omml';   // v1 default: matplotlib
  page_size: 'letter' | 'a4';
  column_sep_pt?: number;                           // column gutter in points; defaults to IEEE spec
}

// ── Root ──────────────────────────────────────────────────────────────────────

export interface DocumentMetadata {
  title: string;
  authors: Author[];
  affiliations: Affiliation[];
  abstract: string;
  keywords: string[];
  conference?: string;          // e.g. "IEEE ICASSP 2025"
}

export interface DocumentModel {
  schema_version: typeof SCHEMA_VERSION;
  metadata: DocumentMetadata;
  blocks: Block[];
  references?: Reference[];
  compile_options: CompileOptions;
}
